pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TopicTransactionBufferRecoverTest.java

```diff
@@ -818,9 +820,98 @@ public void openReadOnlyManagedLedgerFailed(ManagedLedgerException exception, Ob
+                .topic(topic)
+                .subscriptionName(subName)
+                .subscriptionType(SubscriptionType.Exclusive)
+                .subscribe();
+
+        // 2. Check the AbortedTxnProcessor workflow 10 times
+        int messageSize = theSizeOfSegment * 4;
+        for (int i = 0; i < 10; i++) {
+            MessageId maxReadMessage = null;
+            int abortedTxnSize = 0;
+            for (int j = 0; j < messageSize; j++) {
+                Transaction transaction = pulsarClient.newTransaction()
+                        .withTransactionTimeout(5, TimeUnit.MINUTES).build().get();
+                //Half common message and half transaction message.
+                if (j % 2 == 0) {
+                    MessageId messageId = producer.newMessage(transaction).value(i * 10 + j).send();
+                    //And the transaction message have a half which are aborted.
+                    if (RandomUtils.nextInt() % 2 == 0) {
+                        transaction.abort().get();
+                        abortedTxns.add(messageId);
+                        abortedTxnSize++;
+                    } else {
+                        ongoingTxns.put(transaction, messageId);
+                        if (maxReadMessage == null) {
+                            //The except number of the messages that can be read
+                            maxReadMessage = messageId;
+                        }
+                    }
+                } else {
+                    producer.newMessage().value(i * 10 + j).send();
+                    transaction.commit().get();
+                }
+            }
+            // 2.1 Receive all message before the maxReadPosition to verify the correctness of the max read position.
+            int hasReceived = 0;
+            while (true) {
+                Message<Integer> message = consumer.receive(2, TimeUnit.SECONDS);
+                if (message != null) {
+                    Assert.assertTrue(message.getMessageId().compareTo(maxReadMessage) < 0);
+                    hasReceived ++;
+                } else {
+                    break;
+                }
+            }
+            //2.2 Commit all ongoing transaction and verify that the consumer can receive all rest message
+            // expect for aborted txn message.
+            for (Transaction ongoingTxn: ongoingTxns.keySet()) {
+                ongoingTxn.commit().get();
+            }
+            ongoingTxns.clear();
+            for (int k = hasReceived; k < messageSize - abortedTxnSize; k++) {
+                Message<Integer> message = consumer.receive(2, TimeUnit.SECONDS);
+                assertNotNull(message);
+                assertFalse(abortedTxns.contains(message.getMessageId()));
+            }
+        }
+        // 3. After the topic unload, the consumer can receive all the messages in the 10 tests
+        // expect for the aborted transaction messages.
```

Comment:
before unload, check the aborted txn, segment, index stats from the processor, after unload check again. and add the tirm and clear test.

Reply:
There is a clear snapshot test clearTransactionBufferSnapshotTest in TopicTransactionBufferRecoverTest.
And trim test in testPutAbortedTxnIntoProcessor.
```java
  //3. Delete the ledgers and then verify the date.
        Field ledgersField = ManagedLedgerImpl.class.getDeclaredField("ledgers");
        ledgersField.setAccessible(true);
        NavigableMap<Long, MLDataFormats.ManagedLedgerInfo.LedgerInfo> ledgers =
                (NavigableMap<Long, MLDataFormats.ManagedLedgerInfo.LedgerInfo>)
                        ledgersField.get(persistentTopic.getManagedLedger());
        ledgers.forEach((k, v) -> {
            ledgers.remove(k);
        });
```

Reply:
> before unload, check the aborted txn, segment, index stats from the processor
The processor's internal stat verity can be found in `SegmentAbortedTxnProcessorTest`.
This test will verify the processor can run properly end to end.